### PR TITLE
tiqb_query: Support validating arguments in `rpn_expr/test_util.rs`

### PR DESCRIPTION
**Pull request by mapleFU**
## What have you changed? (mandatory)

Coprocessor uses `RpnFnScalarEvaluator` in `rpn_expr/test_util.rs` to testing RPN functions, and previous `RpnFnScalarEvaluator::evaluate` would not use `validator_ptr` in `RpnFnMeta` to checking if the arguments are valid. 

This pr add validation phase in `RpnFnScalarEvaluator::evaluate`, which will checking the arguments before calling Rpn Functions.

## What are the type of changes? (mandatory)

- Engineering (update RPN Function testing tools)

## How has this PR been tested? (mandatory)

No testing now.

## Does this PR affect documentation (docs) or release note? (mandatory)

no

## Does this PR affect `tidb-ansible` update? (mandatory)

No.

### CODE REVIEW

**Comment by breeswish** [comment]
I think it would be more correct to *always* check the validation of the input in `RpnFnScalarEvaluator`, since it is a prerequisite that needs assertion.

**Comment by mapleFU** [comment]
> I think it would be more correct to _always_ check the validity of the input in `RpnFnScalarEvaluator`, since it is a prerequisite of this structure and it needs assertion. Currently invalid inputs do cause test failures but the failures are popped up from a wrong place, which should be corrected.

Yes, I'm now testing if other tests in coprocessor support the `evaluate_with_validate`. If they do, I will apply validation phase to all `RpnFnScalarEvaluator` evaluate.

**Comment by breeswish** [comment]
> Yes, I'm now testing if other tests in coprocessor support the evaluate_with_validate. If they do, I will apply validation phase to all RpnFnScalarEvaluator evaluate.

They *should*. If not, then it is a bug that needs to be fixed since validity is a prerequisite, which means invalid inputs are buggy tests.

**Comment by mapleFU** [comment]
> > Yes, I'm now testing if other tests in coprocessor support the evaluate_with_validate. If they do, I will apply validation phase to all RpnFnScalarEvaluator evaluate.
> 
> They _should_. If not, then it is a bug that needs to be fixed since validity is a prerequisite, which means invalid inputs are buggy tests.

Fortunately ` cargo test -p tidb_query` pass all the tests in `tidb_query`. Validation phase has been added in the latest commit.

**Comment by mapleFU** [comment]
> Why not just use `RpnExpressionBuilder::build_from_expr_tree`?

Do you mean using `RpnExpressionBuilder::build_from_expr_tree` instead of the code below?
```
let expr = self
            .rpn_expr_builder
            .push_fn_call(func, children_ed.len(), ret_field_type)
            .build();
```

I think we already organize the argument in Rpn order, `push_fn_call` and `build` is convenient and efficient here.

**Comment by sticnarf** [comment]
> Why not just use `RpnExpressionBuilder::build_from_expr_tree`?

Never mind. A single unit test should not depend on too many other functions so that we can locate bugs more easily.

File: components/tidb_query/src/rpn_expr/types/test_util.rs [hunk 1]
```diff
@@ -109,19 +100,17 @@ impl RpnFnScalarEvaluator {
...
+        let mut fun_sig_expr = Expr::default();
+        fun_sig_expr.set_sig(sig);
+        fun_sig_expr.set_children(children_ed.clone().into());
+        fun_sig_expr.set_field_type(ret_field_type.clone());
 
-            if let Err(e) = (func.validator_ptr)(&fun_sig_expr) {
-                return (Err(e), context);
-            }
+        if let Err(e) = (func.validator_ptr)(&fun_sig_expr) {
+            return (Err(e), context);
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/types/test_util.rs]: Maybe use assertion is better since we did not check any error body.

**Pull request merged**